add space dev documentation (#55)
deta/space-docs: src/pages/docs/en/basics/local.md

---
title: Developing Locally
section: Basics
position: 3
hidden: false
layout: "@docs"
---

## Starting a Development Server

Once you've created a Space project, you can start developing your app locally using the Space CLI's `space dev` command.

```bash

> space dev --open # start a local development server and open the app in your browser

Reviewer: I would use the most simplest version of the command here and mention the open flag further down:

```suggestion
space dev
```

```

Behind the scenes, the Space CLI is:

- Generating a data key to allow your app to access the Bases and Drives of your project
- Running each of your micro in a separate process
- Exposing all of your micros through a single entrypoint, on `localhost:4200` by default.

You can customize the development command for each micro in your `Spacefile` using the `dev` key.

```yaml
v: 0
micros:
  - name: client
    src: ./client
    engine: svelte
    primary: true
    path: client
  - name: api
    src: ./api
    engine: nodejs16
    dev: nodemon index.js # start the micro in development mode
    path: api
```

> The specified `dev` commands need to start a web server that listens on the port specified with the `PORT` environment variable. This way the CLI can properly forward requests to the micro.

Reviewer: We should link to the instructions for Python and the other engines here

## Advanced Usage

If you want to have more control over the development process, you can also split the `dev` command into two parts:

- First, start micros individually using the `space dev up` command
- Then, start the entrypoint server using the `space dev proxy`

For our example above, this would look like this:

```bash
# Start the micros individually
space dev up client
space dev up api

# Start the entrypoint server
space dev proxy
```

Some use cases for this approach include:

- Running each micro in a separate terminal window
- integrating micros with other development tools (ex: vscode tasks)

The `space dev` command is just a shortcut for the combination of `space dev up` and `space dev proxy`.

## Running scheduled actions manually

> You can trigger scheduled actions from your local development server by using the `space dev trigger` command.

Reviewer: Link to the scheduled actions docs

---
title: Developing Locally
section: Basics
position: 3
hidden: false
layout: "@docs"
---

## Starting a Development Server

> Once you've created a Space project, you can start developing your app locally using the Space CLI's `space dev` command.

Reviewer: ```suggestion
Once you've created a Space project, you can start developing your app locally using the [Space CLI's](https://deta.space/docs/en/reference/cli) `space dev` command.
```

---
title: Developing Locally
section: Basics
position: 3
hidden: false
layout: "@docs"
---

## Starting a Development Server

Once you've created a Space project, you can start developing your app locally using the Space CLI's `space dev` command.

```bash
space dev # start a local development server and open the app in your browser
```

Behind the scenes, the Space CLI is:

- Generating a data key to allow your app to access the Bases and Drives of your project
- Running each of your micro in a separate process
- Exposing all of your micros through a single entrypoint, on `localhost:4200` by default.

> If you want to open the app in your browser, you can use add `--open` flag:
> 
> ```bash
> space dev --open # start a local development server and open the app in your browser
> ```

Reviewer: I think this can be shortened:

````suggestion
If you want to open the app in your browser automatically after starting your Micros, use the `--open` flag.
````

You can customize the development command for each micro in your `Spacefile` using the `dev` key.

```yaml
v: 0
micros:
  - name: client
    src: ./client
    engine: svelte
    primary: true
    path: client
  - name: api
    src: ./api
    engine: nodejs16
    dev: nodemon index.js # start the micro in development mode
    path: api
```

The specified `dev` commands need to start a web server that listens on the port specified with the `PORT` environment variable. This way the CLI can properly forward requests to the micro.

> You can find guide specific to each micro engine in the Quickstart Guides.

Reviewer: ```suggestion
You can find instructions on how to setup the right `dev` command for each engine in the "Quickstart Guides".
```

## Advanced Usage

If you want to have more control over the development process, you can also split the `dev` command into two parts:

- First, start micros individually using the `space dev up` command

> - Then, start the entrypoint server using the `space dev proxy`

Reviewer: "entrypoint server" sounds a bit weird lol. Don't know a better name though except proxy

Author: yeah proxy is fine, it is in the command name

For our example above, this would look like this:

```bash
# Start the micros individually
space dev up client
space dev up api

# Start the entrypoint server
space dev proxy
```

Some use cases for this approach include:

- Running each micro in a separate terminal window
- integrating micros with other development tools (ex: vscode tasks)

The `space dev` command is just a shortcut for the combination of `space dev up` and `space dev proxy`.

## Running scheduled actions manually

You can trigger [scheduled actions](/docs/en/basics/micros#scheduled-actions) from your local development server by using the `space dev trigger` command.

First make sure you have defined a scheduled action in your Spacefile and set up the right handler in your Micro:

> ```yaml

Reviewer: Just a thought: would be cool if we have a component for the docs where you can show different code blocks as "file" tabs. Then we could have a tab for the Spacefile and one for the Micro handler

---
title: Running locally
section: Basics
position: 3
layout: "@docs"
---

## Starting a Development Server

> 
> Once you've created a project, you can start developing your app locally.
> You can use the `space dev` command to start a local development server for your project.

Reviewer: ```suggestion

Once you've created a Space project, you can start developing your app locally using the Space CLI's `space dev` command. The `space dev` command allows you to use Deta Base and Deta Drive with no extra configuration and emulates the [routing of Deta Space](/docs/en/basics/micros#micro-routing), specified in your Spacefile.
```

> 
> ```bash
> space dev --open # start a local development server and open the app in your browser
> ```
> 
> You can control what command is run for each micro by editing the `dev` field in your `Spacefile`.
> 
> ```yaml
> v: 0
> micros:
>   - name: client
>     src: ./client
>     engine: svelte
>     primary: true
>     path: client
>   - name: api
>     src: ./api
>     engine: nodejs16
>     dev: node index.js
>     path: api
> ```
> 

Reviewer: ````suggestion

```bash
space dev --open # start a local development server and open the app in your browser
```

When running `space dev` the CLI will start its own web server on port `4200` that emulates the Deta Space runtime locally. It will start each Micro by running the `dev` command that is either set by the engine automatically or specified in the `Spacefile`. It automatically handles the routing between Micros and logs from every Micro are combined and sent to the console.

You can control what command is run for each Micro by editing the `dev` field in your `Spacefile`.

```yaml
v: 0
micros:
  - name: client
    src: ./client
    engine: svelte
    primary: true
    path: client
  - name: api
    src: ./api
    engine: nodejs16
    dev: node index.js
    path: api
```

The specified `dev` commands need to start a web server that listens on the port specified with the `PORT` environment variable. This way the CLI can properly forward requests to the Micro.

To develop and test your app open a browser and navigate to `http://localhost:4200`.

````

## Running micros individually

Optionally you can instead use separate commands to run each micro individually, then use a reverse proxy to route requests to the correct micro.

> 

Reviewer: This is very unclear and confusing.

Author: you are right, I tried to reuse the example above to make it clearer.

```bash
space dev up client # start client
space dev up api # start api

space dev proxy --port 3000 # start a reverse proxy on port 3000 in front of all micros
```

> **Note** The `space dev` command only takes care of running your micros locally.
> You will need to install all micro dependencies before running the `space dev` command.

## Triggering Actions

> You can trigger actions from your local development server by using the `space dev trigger` command.

Reviewer: Needs more explanation as well

Author: done in 7e4cece

```bash
space dev up backend # start your backend micro

space dev trigger action-name # trigger an action (the action's micro must be running)
```

## Running commands

> If you want to run a single command in the context of your project, you can use the `space run` command.

Reviewer: Might be good to mention some use cases for this i.e. one of scripts, automating importing/exporting data from your project etc.
